java/src/jni/h5util.c

```diff
@@ -4125,64 +4641,50 @@ translate_rbuf(JNIEnv *env, jobjectArray ret_buf, jlong mem_type_id, H5T_class_t
-
-            if (nmembs < 0)
-                goto done;
-
-            /* Convert each element to a list */
-            for (i = 0; i < (size_t)nmembs; i++) {
-                jboolean     found_jList = JNI_TRUE;
-                jobjectArray jList       = NULL;
-                int          ret_buflen  = -1;
-
-                if ((memb = H5Tget_member_type(mem_type_id, (unsigned int)i)) < 0)
-                    H5_LIBRARY_ERROR(ENVONLY);
-                offset = H5Tget_member_offset(mem_type_id, (unsigned int)i);
+            void *objBuf = NULL;
 
-                if ((vlClass = H5Tget_class(memb)) < 0)
-                    H5_LIBRARY_ERROR(ENVONLY);
-                if (!(vlSize = H5Tget_size(memb)))
-                    H5_LIBRARY_ERROR(ENVONLY);
+            /* Convert each compound element to a list */
+            for (i = 0; i < (size_t)count; i++) {
+                /* The list we're going to return: */
+                if (NULL == (jList = (jobjectArray)ENVPTR->NewObject(ENVONLY, arrCList, arrListMethod, 0)))
+                    H5_OUT_OF_MEMORY_ERROR(ENVONLY, "translate_rbuf: failed to allocate list read buffer");
+                int nmembs = H5Tget_nmembers(mem_type_id);
+                /* Convert each element to a list */
+                for (x = 0; x < (size_t)nmembs; x++) {
+                    H5T_class_t memb_vlClass;
+                    size_t      memb_vlSize;
+                    size_t      memb_offset;
+
+                    if ((memb = H5Tget_member_type(mem_type_id, (unsigned int)x)) < 0)
+                        H5_LIBRARY_ERROR(ENVONLY);
+                    if ((memb_offset = H5Tget_member_offset(mem_type_id, (unsigned int)x)) < 0)
+                        H5_LIBRARY_ERROR(ENVONLY);
 
-                /* Get the object element */
-                HDmemcpy(&objBuf, (char *)raw_buf + offset, vlSize);
+                    if ((memb_vlClass = H5Tget_class(memb)) < 0)
+                        H5_LIBRARY_ERROR(ENVONLY);
+                    if (!(memb_vlSize = H5Tget_size(memb)))
+                        H5_LIBRARY_ERROR(ENVONLY);
 
-                ret_buflen = ENVPTR->GetArrayLength(ENVONLY, ret_buf);
-                /* The list we're going to return: */
-                if (i < ret_buflen) {
-                    if (NULL ==
-                        (jList = ENVPTR->GetObjectArrayElement(ENVONLY, (jobjectArray)ret_buf, (jsize)i)))
-                        found_jList = JNI_FALSE;
-                }
-                if (NULL == jList) {
-                    if (NULL ==
-                        (jList = (jobjectArray)ENVPTR->NewObject(ENVONLY, arrCList, arrListMethod, 0)))
-                        H5_OUT_OF_MEMORY_ERROR(ENVONLY,
-                                               "translate_rbuf: failed to allocate list read buffer");
+                    translate_atomic_rbuf(ENVONLY, jList, memb, memb_vlClass,
+                                          ((char *)raw_buf) + i * typeSize + memb_offset);
```

Comment:
More of a nitpick since I'm assuming this works correctly as is, but for "container" types like compound, vlen, etc. it would probably be more efficient to check whether each member is also a container type and pass them directly back to translate_rbuf/wbuf rather than passing to translate_atomic_rbuf/wbuf just to have those pass back up to translate_rbuf/wbuf. It would also make more sense considering the "atomic" part of the functions names, but this is mostly just a design issue.

Reply:
Working on a solution for non-ArrayList objects as data (next PR soon) which may address this.